### PR TITLE
Adding drone as CI

### PR DESCRIPTION
**Pull request by nupplaphil**
This is an outlook how we can use Drone CI instead of Travis for an upcoming migration to Gitea (where we cannot use Travis anymore).

The tests are currently starting on my node:
https://drone.opensocial.at/nupplaphil/friendica

ToDo:
- [x] Make DB working (don't know why the dataset isn't loaded)
- [x] Skip unavailable caching mechanism
- [x] general overview
- [x] Make CodeCov work again
- [x] Setup MySQL and MariaDB tests
- [x] Setup extra Redis, Memcache, Memcached tests (fast and don't need a DB)
- [x] Switch codecov Integration from travis to drone (@tobiasd needed :) )
- [x] Setup Drone instance https://drone.friendi.ca for CI (@ben-utzer needed for setup and @tobiasd for the OAuth token)
- [x] Add a doc about `autotest.sh` because it can be used during development too

### CODE REVIEW

**Comment by nupplaphil** [comment]
Erm... I accidentaly doubled the speed of the travis tests because of phpunit.xml restructuring.. I've to investigate what happend and cherry pick it for RC too..

File: .drone.yml [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+steps:
+- name: mysql-php7.1
+  image: friendicaci/php7.1:php7.1.32
+  commands:
+      - phpenmod xdebug
+      - sleep 20
+      - ./autotest.sh
+      - wget https://codecov.io/bash -O codecov.sh
+      - sh -c "if [ '$DRONE_BUILD_EVENT' = 'pull_request' ]; then bash codecov.sh -B $DRONE_BRANCH -C $DRONE_COMMIT -P $DRONE_PULL_REQUEST -t 2f4b253b-ca17-41d7-96e3-81623581c97d -f tests/autotest-clover.xml; fi"
+      - sh -c "if [ '$DRONE_BUILD_EVENT' != 'pull_request' ]; then bash codecov.sh -B $DRONE_BRANCH -C $DRONE_COMMIT -t 2f4b253b-ca17-41d7-96e3-81623581c97d -f tests/autotest-clover.xml; fi"
```
Comment (MrPetovan) [inline .drone.yml]: Can't you put this in an `else` statement in the previous line?

File: bin/wait-for-connection [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	if ($sock !== false) {
+		fclose($sock);
+		fwrite(STDOUT, "\n");
+		exit(0);
+	}
+	sleep(1);
+	fwrite(STDOUT, '.');
+} while (time() < $stopTime);
+fwrite(STDOUT, "\n");
+exit(1);
```
Comment (MrPetovan) [inline bin/wait-for-connection]: Code standards: Please add an EOF EOL.

File: src/Core/Cache/MemcacheGetAllKeys.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+			reset($a_slab);
+			foreach ($a_slab AS &$a_key) {
+				$keys[] = $a_key;
+			}
+		}
+		unset($slab);
+
+		return $keys;
+	}
+}
```
Comment (MrPetovan) [inline src/Core/Cache/MemcacheGetAllKeys.php]: Code standards: Please add an EOF EOL.

**Comment by nupplaphil** [comment]
@mrpetovan only memcache, but the getallkeys method is now a (necessary) real ugly workaround for both. Local the tests are working, but on drone, there are some strange issues for both memcache classes now.. Seems like a race condition, but I'm not sure. Shall I cherry pick this bugfix to the RC branch?

**Comment by MrPetovan** [comment]
> Shall I cherry pick this bugfix to the RC branch?

We haven't had any user report about any Memcache(d) issue so I would say it can wait until next version.

**Comment by nupplaphil** [comment]
:P ... Locally the workaround is working and all tests are passing for both memcache methods, but on Drone, the workaround seems to not work at all ... And I've currently no clue why ^^

**Comment by nupplaphil** [comment]
@tobiasd can you login at https://codecov.io/gh/friendica/friendica with the Friendica Github account and at settings, there's a field "Token". This token is needed for the drone-config to upload the drone results to codecov :) so plz send me the token and I'll add it in this PR :(

Thanks

**Comment by nupplaphil** [comment]
At least.. All lights are green :) .. There's some room for improvement, but for now I'm finished. Next step ist the drone instance :)

File: src/Core/Cache/TraitMemcacheCommand.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		$lines  = explode("\r\n", $string);
+		$slabs  = [];
+		$keys   = [];
+
+		foreach ($lines as $line) {
+
+			if (preg_match("/STAT items:([\d]+):number ([\d]+)/", $line, $matches)) {
+
+				if (isset($matches[1])) {
+					if (!in_array($matches[1], $keys)) {
```
Comment (MrPetovan) [inline src/Core/Cache/TraitMemcacheCommand.php]: Please combine the three conditional clauses into one.

File: src/Core/Cache/TraitMemcacheCommand.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	{
+		$s = @fsockopen($this->server, $this->port);
+		if (!$s) {
+			throw new InternalServerErrorException("Cant connect to:" . $this->server . ':' . $this->port);
+		}
+
+		fwrite($s, $command . "\r\n");
+		$buf = '';
+
+		while ((!feof($s))) {
```
Comment (MrPetovan) [inline src/Core/Cache/TraitMemcacheCommand.php]: Please remove the extraneous pair of parentheses.

**Comment by nupplaphil** [comment]
Improved caching :) the DB is now running inside tmpfs volumes which increases the overall speed. We're now almost as fast as travis, but we're additionally testing MariaDB.
In the future, it will now be easier to add more test-environments (e.g. Postgres, Sqlite tests or UI-tests ;) )

File: .codecov.yml [hunk 6]
```diff
@@ -1,10 +1,15 @@
+codecov:
+  branch: develop
+  ci:
+    - drone.friendi.ca
+    - !scrutinizer-ci.com
```
Comment (MrPetovan) [inline .codecov.yml]: What is this for?

**Comment by nupplaphil** [comment]
Works :-)
Adapt codecov as well

=> I'm finished and ready for merging

File: .codecov.yml (hunk 6 above)
Comment (nupplaphil) [reply .codecov.yml]: copy & paste fail ;-)

**Comment by nupplaphil** [comment]
travis is obviously fast now because I disabled the codecov export, which took the most of the time (creating an XML over the whole coverage of the codebase)

File: .travis.yml [hunk 7]
```diff
@@ -26,4 +26,4 @@ before_script:
  - phpenv config-add .travis/redis.ini
  - phpenv config-add .travis/memcached.ini
 
-after_success: bash <(curl -s https://codecov.io/bash)
+script: vendor/bin/phpunit --configuration tests/phpunit.xml
```
Comment (MrPetovan) [inline .travis.yml]: Please add an EOF EOL

File: autotest.sh [hunk 8]
```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/env bash
+
```
Comment (MrPetovan) [inline autotest.sh]: Please add a comment explaining the goal and usage of this script.

File: bin/wait-for-connection [hunk 9]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/php
+
```
Comment (MrPetovan) [inline bin/wait-for-connection]: Please add a comment explaining the goal and usage of this script.

**Pull request merged**